src/shogun/machine/EnsembleMachine.h

```diff
@@ -53,10 +52,36 @@ namespace shogun
 		    const std::shared_ptr<Features>& data,
 		    const std::shared_ptr<Labels>& labs)
 		{
-			for (auto&& machine : m_machines)
+			const int32_t& num_threads = env()->get_num_threads();
+			if (num_threads > 1)
 			{
-				machine->set_labels(labs);
-				machine->train(data);
+				std::vector<std::thread> threads;
+				int32_t i = 0, num_machine = m_machines.size();
+				for (i = 0; i < std::min(num_threads, num_machine); i++)
+				{
+					const auto& machine = m_machines[i];
+					threads.emplace_back([&]() {
+						machine->set_labels(labs);
+						machine->train(data);
+					});
+				}
+				for (auto&& thread : threads)
+				{
+					thread.join();
+				}
+				for (; i < m_machines.size(); i++)
```

Comment:
but then if I have 100 machines and only set 8 threads, 92 machines will be trained in sequentially?